src/Components/Main/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,161 @@
+import {
+  memo,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import cn from 'classnames';
+import { useTodoContext } from '../../../Context/Context';
+import { Todo } from '../../../types/Todo';
+import { TemporaryTodo } from '../TemporaryTodo/TemporaryTodo';
+
+export const TodoItem = memo(() => {
+  const [todoForEdit, setTodoForEdit] = useState<Todo | null>(null);
+  const [editTitle, setEditTitle] = useState('');
+  const {
+    filteredList,
+    tempTodo,
+    handleDelete,
+    onDelete,
+    editTodo,
+    toggleTodoCompleted,
+    setOnDelete,
+    errorMessage,
+  } = useTodoContext();
+
+  const titleInput = useRef<HTMLInputElement>(null);
+  const escapeFunction = useCallback((event:KeyboardEvent) => {
+    if (event.key === 'Escape') {
+      setTodoForEdit(null);
+    }
+  }, []);
+
+  const handleTitleForm = useCallback((event:React.FormEvent) => {
+    event.preventDefault();
+
+    if (!todoForEdit) {
+      return;
+    }
+
+    if (todoForEdit.title === editTitle.trim()) {
+      setTodoForEdit(null);
+
+      return;
+    }
+
+    if (editTitle === '') {
+      handleDelete(todoForEdit.id).then(() => {
+        setTodoForEdit(null);
+      });
+
+      return;
+    }
+
+    editTodo({ ...todoForEdit, title: editTitle.trim() })
+      .then(() => {
+        setTodoForEdit(null);
+      })
+      .catch(() => {})
+      .finally(() => {
+        setOnDelete([]);
+      });
+  }, [editTitle, editTodo, handleDelete,
+    setOnDelete, todoForEdit]);
+
+  useEffect(() => {
+    if (titleInput.current) {
+      titleInput.current.focus();
+    }
+
+    document.addEventListener('keyup', escapeFunction, false);
+
+    return () => {
+      document.removeEventListener('keyup', escapeFunction, false);
+    };
+  }, [todoForEdit, escapeFunction, errorMessage, editTitle]);
+
+  return (
+    <section className="todoapp__main" data-cy="TodoList">
+      {filteredList.map(({
+        title,
+        completed,
+        id,
+        userId,
+      }) => (
+        <div
+          key={id}
+          data-cy="Todo"
+          className={cn('todo',
+            { completed })}
+        >
+          <label className="todo__status-label">
+            <input
+              data-cy="TodoStatus"
+              type="checkbox"
+              className="todo__status"
+              onChange={() => toggleTodoCompleted({
+                title,
+                completed,
+                id,
+                userId,
+              })}
+              checked={completed}
+            />
```

Comment:
In order not to duplicate the code, it is better to do the destructuring separately
```suggestion
      {filteredList.map((todo) => {
      const { title, completed, id, userId } = todo;
      return <div
          key={id}
          data-cy="Todo"
          className={cn('todo', { completed })}
        >
          <label className="todo__status-label">
            <input
              data-cy="TodoStatus"
              type="checkbox"
              className="todo__status"
              onChange={() => toggleTodoCompleted(todo)}
              checked={completed}
            />
```